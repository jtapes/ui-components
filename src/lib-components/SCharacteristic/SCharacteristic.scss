@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-characteristic {
  display: inline-flex;
  align-items: center;

  & + & {
    margin-left: $size--16;
  }
}
