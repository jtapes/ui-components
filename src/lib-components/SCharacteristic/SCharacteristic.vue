<template>
  <div class="s-characteristic" :data-test="name">
      <svg-icon v-if="isShowIcon" :name="`${nameIcon}-16`" class="icon-16 mr-4" />
    <span class="h4">{{ value }}</span>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'

export const CHARACTERISTICS = [
  'thickness',
  'width',
  'length',
  'diameter',
  'channel',
  'color',
  'thickness-cover',
]

/** Компонент для отображения основных признаков товара */
@Component({
  components: {
    SvgIcon,
  }
})
export default class SCharacteristic extends Vue {
  /** Название иконки признака */
  @Prop({
    type: String,
    validator(value: string): boolean {
        return CHARACTERISTICS.includes(value) || !(value)
    },
  })
  nameIcon!: string

  /** Наименование признака */
  @Prop({ type: String, required: true })
  name!: string

  /** Значение признака */
  @Prop([String, Number])
  value!: string | number

  get isShowIcon(): boolean {
    return CHARACTERISTICS.includes(this.nameIcon)
  }
}
</script>
