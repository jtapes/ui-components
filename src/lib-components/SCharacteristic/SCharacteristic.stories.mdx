import { config, withDesign } from 'storybook-addon-designs'

import { withTests } from '@storybook/addon-jest';

import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Canvas,
} from '@storybook/addon-docs'

import results from '/.jest-test-results.json';

import SCharacteristic from '@/lib-components/SCharacteristic/SCharacteristic.vue'

<Meta
  title="DESIGN SYSTEM/Market/Characteristic"
  component={SCharacteristic}
  decorators={[withDesign, withTests({results})]}
  argTypes={{
    nameIcon: {
      options: [
        'thickness',
        'width',
        'length',
        'diameter',
        'channel',
        'color',
        'thickness-cover',
      ],
      control: { type: 'select' },
    },
    name: { control: 'text' },
    value: { control: 'text' },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/42Xh2qI3qvq2CRkc2HJlwA/MVP-3-Market-Catalog-%E2%80%A2-Product-Page?node-id=1150%3A120',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Основной признак товара

<Description of={SCharacteristic} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SCharacteristic },
  template:
    `<s-characteristic :nameIcon="nameIcon" :name="name" :value="value" />`,
})

<Canvas withToolbar withSource="none">
  <Story name="SCharacteristic" args={{ name: 'length', value: 2000 }}>
    {Template.bind({})}
  </Story>
</Canvas>

## Пример

<Source
  language="html"
  code={`<s-characteristic :nameIcon="length" name="length" value="2000" />`}
/>

<ArgsTable of={SCharacteristic} />
