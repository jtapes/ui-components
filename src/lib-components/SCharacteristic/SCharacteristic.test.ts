import { mount } from '@vue/test-utils'
import SCharacteristic from '@/lib-components/SCharacteristic/SCharacteristic.vue'

describe('SCharacteristic', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SCharacteristic, {
      propsData: {
        nameIcon: 'length',
        name: 'length',
        value: 2000,
      },
    })

    expect(wrapper.element).toMatchSnapshot()
  })
  test('should be "console.log" without "name" prop', () => {
    const consoleWarnMock = jest.spyOn(console, 'error').mockImplementation()
    const wrapper = mount(SCharacteristic, {})
    expect(consoleWarnMock).toHaveBeenCalledTimes(1)
    expect(wrapper.text()).toBe('')
    consoleWarnMock.mockRestore()
  })

  test.each(['',null,undefined,'any_text'])
  ('should be without "svg" when prop nameIcon="%s"', (a) => {
    const wrapper = mount(SCharacteristic, {
      propsData: {
        nameIcon: a,
        name: 'length',
      },
    })
    expect(wrapper.find('svg').exists()).toBeFalsy()
    wrapper.destroy()
  })

  test('should be "name" characteristics without icon', () => {
    const value = '2000'
    const wrapper = mount(SCharacteristic, {
      propsData: {
        name: 'length',
        value,
      },
    })
    expect(wrapper.find('svg').exists()).toBeFalsy()
    expect(wrapper.text()).toBe(value)
  })

  test('should be "name" characteristics with icon and special class', () => {
    const nameIcon = 'length'
    const wrapper = mount(SCharacteristic, {
      propsData: {
        nameIcon,
        name: 'length',
        value: '3000',
      },
    })
    expect(wrapper.find('svg').exists()).toBeTruthy()
    expect(wrapper.find('svg').classes().join()).toContain(nameIcon)
  })

})
