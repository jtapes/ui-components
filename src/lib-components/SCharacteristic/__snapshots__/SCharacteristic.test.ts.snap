// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SCharacteristic Is Vue snapshot 1`] = `
<div
  class="s-characteristic"
  data-test="length"
>
  <svg
    class="svg-icon icon-16 mr-4 svg-icon--length-16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#length-16"
      xlink:href="#length-16"
    />
  </svg>
   
  <span
    class="h4"
  >
    2000
  </span>
</div>
`;
